File: packages/ts/src/validate.ts
import type { Deps, Field, TasksDuck } from './types'

import { throwError } from './utils'

export function validateDeps<Tasks extends TasksDuck>(
  fieldMap: Record<Field<Tasks>, true>,
  deps: Deps<Tasks>,
) {
  Object.keys(deps).forEach((field) => {
    if (!fieldMap[field]) {
      throwError({
        title: 'validateDeps',
        msg: `deps[${field}] is not defined in tasks`,
      })
    }
    deps[field]?.some((depItem) => {
      if (depItem === field) {
        throwError({
          title: 'validateDeps',
          msg: `deps[${field}]${depItem} must not be the same`,
        })
      }
      if (!fieldMap[depItem]) {
        throwError({
          title: 'validateDeps',
          msg: `deps[${field}]${depItem} is not defined in tasks`,
        })
      }
      return true
    })
  })
}

export function validateTasks<Tasks extends TasksDuck>(tasks: Tasks) {
  Object.keys(tasks).forEach((field) => {
    const task = tasks[field]
    if (typeof task !== 'function') {
      throwError({
        title: 'validateTasks',
        msg: `task[${field}] must be a function`,
      })
    }
  })
}

export function validateRunningTasks<Field extends keyof any>(
  fieldMap: Record<Field, true>,
  fields: Field[],
) {
  fields.forEach((field) => {
    if (!fieldMap[field]) {
      throwError({
        title: 'validateRunningTasks',
        msg: `${field as string} is not defined in tasks`,
      })
    }
  })
}

File: core/validate.go
package core

import "fmt"

func ValidateDeps(fieldMap map[string]bool, deps map[string][]string) error {
  for field, depFields := range deps {
    if !fieldMap[field] {
      message := fmt.Sprintf("deps[%s] is not defined in tasks", field)
      return FormatError("validateDeps", message)
    }
    for _, depItem := range depFields {
      if depItem == field {
        message := fmt.Sprintf("deps[%s] has a circular dependency", field)
        return FormatError("validateDeps", message)
      }
      if !fieldMap[depItem] {
        message := fmt.Sprintf("deps[%s]%s is not defined in tasks", field, depItem)
        return FormatError("validateDeps", message)
      }
    }
  }
  return nil
}

func ValidateRunningTasks(fieldMap map[string]bool, fields []string) error {
  for _, field := range fields {
    if !fieldMap[field] {
      message := fmt.Sprintf("%s is not defined in tasks", field)
      return FormatError("validateRunningTasks", message)
    }
  }
  return nil
}
